### PR TITLE
⚠️ Support bootstrap data in secret

### PR DESCRIPTION
**Pull request by vincepri**
**What this PR does / why we need it**:
This PR is a follow-up requirement to an upstream change that now stores the bootstrap data in a secret rather than inline plaintext field.

### CODE REVIEW

File: pkg/cloud/scope/machine.go [hunk 1]
```diff
@@ -183,6 +184,21 @@ func (m *MachineScope) SetAddresses(addrs []corev1.NodeAddress) {
...
+// GetBootstrapData returns the bootstrap data from the secret in the Machine's bootstrap.dataSecretName.
+func (m *MachineScope) GetBootstrapData() (string, error) {
+	if m.Machine.Spec.Bootstrap.DataSecretName == nil {
+		return "", errors.New("error retrieving bootstrap data: linked Machine's bootstrap.dataSecretName is nil")
+	}
+
+	secret := &corev1.Secret{}
+	key := types.NamespacedName{Namespace: m.Namespace(), Name: *m.Machine.Spec.Bootstrap.DataSecretName}
+	if err := m.client.Get(context.TODO(), key, secret); err != nil {
+		return "", errors.Wrapf(err, "failed to retrieve bootstrapd data secret for AWSMachine %s/%s", m.Namespace(), m.Name())
```
Comment (ncdc) [inline pkg/cloud/scope/machine.go]:
```suggestion
		return "", errors.Wrapf(err, "failed to retrieve bootstrap data secret for AWSMachine %s/%s", m.Namespace(), m.Name())
```

Comment (vincepri) [reply pkg/cloud/scope/machine.go]: Fixed

File: pkg/cloud/services/ec2/instances.go [hunk 2]
```diff
@@ -178,7 +179,11 @@ func (s *Service) CreateInstance(scope *scope.MachineScope) (*infrav1.Instance,
 	}
 
 	// Set userdata.
-	input.UserData = aws.String(*scope.Machine.Spec.Bootstrap.Data)
+	userData, err := scope.GetBootstrapData()
+	if err != nil {
+		return nil, err
```
Comment (ncdc) [inline pkg/cloud/services/ec2/instances.go]: Sorry for the additional comment, but I think this error will get "swallowed" in that it's only written to the pod log, and it's never visible to the user anywhere. We generate events in this method (CreateInstance), but only for certain failures. Maybe we should merge this as-is, and have a follow-up issue to adjust the success/fail event creation in CreateInstance so it covers all returnable errors?

File: pkg/cloud/scope/machine.go [hunk 3]
```diff
@@ -183,6 +184,21 @@ func (m *MachineScope) SetAddresses(addrs []corev1.NodeAddress) {
 	m.AWSMachine.Status.Addresses = addrs
 }
 
+// GetBootstrapData returns the bootstrap data from the secret in the Machine's bootstrap.dataSecretName.
+func (m *MachineScope) GetBootstrapData() (string, error) {
+	if m.Machine.Spec.Bootstrap.DataSecretName == nil {
+		return "", errors.New("error retrieving bootstrap data: linked Machine's bootstrap.dataSecretName is nil")
```
Comment (ncdc) [inline pkg/cloud/scope/machine.go]: This error should probably include the machine's ns & name?

File: pkg/cloud/services/ec2/instances.go (hunk 2 above)
Comment (vincepri) [reply pkg/cloud/services/ec2/instances.go]: Added a error and success event for creation

File: pkg/cloud/scope/machine.go (hunk 3 above)
Comment (vincepri) [reply pkg/cloud/scope/machine.go]: I added the AWSMachine namespace and name from the caller, so all the errors will get it

**Pull request merged**